concept/type_/constraint.rs

```diff
@@ -0,0 +1,144 @@
+/*
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at https://mozilla.org/MPL/2.0/.
+ */
+
+use std::collections::{HashMap, HashSet};
+
+use storage::snapshot::{ReadableSnapshot, WritableSnapshot};
+
+use crate::{
+    error::ConceptReadError,
+    thing::thing_manager::ThingManager,
+    type_::{
+        annotation::{
+            Annotation, AnnotationAbstract, AnnotationCardinality, AnnotationCascade, AnnotationCategory,
+            AnnotationDistinct, AnnotationError, AnnotationIndependent, AnnotationKey, AnnotationRange,
+            AnnotationRegex, AnnotationUnique, AnnotationValues, DefaultFrom,
+        },
+        owns::Owns,
+        type_manager::{type_reader::TypeReader, TypeManager},
+        Capability,
+    },
+};
+
+pub struct Constraint {}
+
+macro_rules! compute_constraint_one_to_one_annotation {
+    ($func_name:ident, $annotation_enum:path, $annotation_type:ident) => {
+        pub fn $func_name<'a, A: Into<Annotation> + Clone + 'a>(
+            annotations: impl IntoIterator<Item = &'a A>,
+        ) -> Option<$annotation_type> {
+            annotations
+                .into_iter()
+                .filter_map(|annotation| match annotation.clone().into() {
+                    $annotation_enum(annotation) => Some(annotation),
+                    _ => None,
+                })
+                .next()
+        }
+    };
+}
+
+#[derive(Debug, Copy, Clone, Eq, PartialEq, Hash)]
+pub enum ConstraintValidationMode {
```

Comment:
My algorithm of annotations revalidation based on schema changes (you'll see it in type_manager, but it's a good place to announce it):
1. Collect all the changed types/capabilities and their annotations (set annotation, changed supertype, changed override, created a new capability for type, unset override (get default cardinality), ....). It may be just one type/capability with one annotation, it may be a list (type/capability) of lists (annotations).
2. For types, just iterate over your instances and instances of your subtypes if they are affected (annotation is inherited). For capabilities, iterate over every pair of {object_type, interface_type} affected (considering overrides, considering inheritance of capability, blablabla...)
3. Just validate the updated types/capabilities, iterating over their instances.

However, steps 2-3 can look different, as we may want to check instances of:
1. Purely one isolated `Type` (regex, values, range): grab one type/cap, iterate over it, validate. Grab its subtype/overriding, iterate over it, validate. Repeat...
2. `Type And Siblings` (cardinality). By `siblings` I mean "we have a common parent (not specifically declared) and we are related to each other in any other way (for capabilities it means being owned by a single type). Grab instances of a type/capability and its siblings and check them together.
3. `Type And Siblings And Subtypes` (unique): collect the whole family together, let them sit in one car. Start the data iteration show only when everyone is in place.

I use this enum for branching between these cases based on specific annotations updated. If I have `card`, `range`, and `regex` updated in one operation, I'll call two instances validations: `Type` and `TypeAndSiblings`. And I use a single function (with some minor preparations) for the algorithm described above, so we can reuse it everywhere (just like I did yesterday when new `card` issues appeared) and append new annotations to it just to several places without worrying about anything else!